Implement deep linking into workflow instance subgraphs

sfmskywalker opened the pull request:
This PR fixes an issue where deep linking into a workflow instance's subgraphs failed due to the new lazy loading implementation.

Fixes #234

sfmskywalker committed:
Revert ActivityWrapper optimized rendering

When the activity is dropped onto the canvas and is embedded inside the MudBadge component, the height calculation is off.

We might be able to fix that, but I'm not sure if it's worth it.

sfmskywalker committed:
Refactor workflow API and models

Reorganize models and API endpoints for workflow path segments. Introduced new `ActivityGraph` record for better activity node handling. Updated `WorkflowInstanceDesigner` and related components to utilize new structure for path segment loading.

sfmskywalker committed:
Refactor activity graph handling and event callback pattern

Replaced the `ActivityGraph` usage with a `RootActivity` property and updated to EventCallback pattern for better performance and code readability. The changes streamline the management of activity graphs and improve the efficiency of event handling through the use of EventCallback, which also enhances maintainability in the codebase.

sfmskywalker committed:
Simplify activity selection logic by using node ID

Refactored the activity selection process to use node ID directly instead of an activity node object. This change ensures a more streamlined and efficient selection process by removing redundant checks and operations. Updated related path segment handling to align with the new method.

sfmskywalker committed:
Remove unused code and references

This commit simplifies the codebase by eliminating unused code and redundant using directives across several components. These changes improve readability and maintainability without affecting existing functionality.

sfmskywalker committed:
Validate and simplify activity handling in flowcharts

Add a type check for "Elsa.Flowchart" before loading the flowchart. Simplify activity merging and embedded activity handling in diagram designer components.

sfmskywalker committed:
Remove commented-out debug code in DiagramDesignerWrapper

Eliminated several lines of commented-out code that were used for debugging purposes. This helps to clean up the file and maintain a clearer codebase.

sfmskywalker committed:
Fix workflow definition container null reference issue

Refactored the container assignment logic to handle cases where no leaf segment is found, ensuring it defaults to `childNode` if ancestors are empty. Additionally, added a TODO comment in DiagramDesignerWrapper.razor.cs for loading activity nodes if not already loaded.

sfmskywalker committed:
Improve activity selection and journal log integration

Refactored workflow instance designer to use local journal log selection instead of a parameter. Enhanced activity selection methods to streamline integration with journal entries. Fixed breadcrumb loading mechanism and adjusted path segment handling for better UX.

sfmskywalker committed:
Exclude Workflow and Flowchart journal entries

Added an exclusion filter for certain activity types when fetching journal logs.

sfmskywalker committed:
Update DiagramDesignerWrapper

Added the ability to first find the selected activity context within the existing graph, before reaching out to the backend.

sfmskywalker committed:
Add activity stats refresh in DiagramDesignerWrapper

Introduced a method to refresh activity statistics when updating breadcrumb items. This enhancement ensures that activity status is current and accurately reflected in the diagram designer.

sfmskywalker committed:
Fix activity graph merge in DiagramDesignerWrapper

Added missing logic to merge nodes into the activity graph after visiting an embedded activity. This ensures that the graph is properly updated when new activities are added.

sfmskywalker committed:
Simplify node path selection logic

Refactor node path selection by removing the redundant logic for finding owning activities, using a backend service for path segments retrieval. This reduces complexity and ensures a consistent approach across different diagram types.

sfmskywalker commented on Elsa.Studio-Core.sln:
```diff
@@ -83,7 +83,7 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "BlazorApp1", "samples\Blazo
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "client", "client", "{F4B408E1-5169-4FA9-B2D7-2377634C4DEE}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Elsa.Api.Client", "..\..\elsa-core\main\src\clients\Elsa.Api.Client\Elsa.Api.Client.csproj", "{26ECE061-0F39-4945-8356-EF14E67A8F39}"
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Elsa.Api.Client", "..\..\elsa-core\3.2.x\src\clients\Elsa.Api.Client\Elsa.Api.Client.csproj", "{26ECE061-0F39-4945-8356-EF14E67A8F39}"
```
Will be updated once the elsa-core PR has been merged.

sfmskywalker commented on src/framework/Elsa.Studio.Core/Elsa.Studio.Core.csproj:
```diff
@@ -33,12 +33,12 @@
 
+    <!--    <ItemGroup Label="Elsa">-->
+    <!--        <PackageReference Include="Elsa.Api.Client"/>-->
+    <!--    </ItemGroup>-->
```
Will be updated once the elsa-core PR has been merged.

sfmskywalker committed:
Streamline workflow editor component and add `radzen` to user dictionary

Changed `IsReadOnly` from public to private in `WorkflowDefinitionEditor.razor` and cleaned up commented-out code in `WorkflowEditor.razor.cs`. Also added the term `radzen` to the user dictionary in the DotSettings file. These changes streamline the workflow editor component and maintain project-specific terminology.

sfmskywalker committed:
Simplify ActivityExecutionsTab.razor.cs by removing comments

Removed redundicative XML comments from ActivityExecutionsTab.razor.cs to improve readability and maintainability of the code. The functionality of the component remains unchanged.

sfmskywalker committed:
Refactor workflow components and remove redundant code

Simplified several workflow-related components by removing redundant and commented-out code. Changed method signatures where appropriate and ensured consistency in code formatting for better readability and maintainability.

sfmskywalker committed:
Remove redundant StateHasChanged call in WorkflowInstanceViewer.

The call to StateHasChanged was unnecessary as the UI updates automatically when the asynchronous operation completes. This change improves performance by avoiding redundant state updates.

sfmskywalker committed:
Remove unused dependency for ActivityVisitor

Removed the unused `IActivityVisitor` dependency from `FallbackDiagramDesigner.cs` to clean up the code. This results in more maintainable and less cluttered code.

raymonddenhaan commented on src/modules/Elsa.Studio.Workflows.Core/Extensions/ActivityVisitorExtensions.cs:
```diff
@@ -4,45 +4,51 @@
     }
     
-    /// <summary>
+    /// Creates an activity graph based on the provided JSON activity.
```
JSON activity -> workflowDefinition

sfmskywalker committed:
Update XML doc to reflect workflow definition

Changed the documentation comment to specify that the activity graph is created based on the provided workflow definition instead of JSON activity. This clarifies the function's purpose and ensures more accurate information.

sfmskywalker committed:
Switch to package reference for Elsa.Api.Client

Updated references to use the latest version of Elsa.Api.Client package instead of project references. This improves dependency management and simplifies updating the client library in future releases.
